File: source/haskell/src/Loops.hs
-- Loops algorithm for cycle counting
{-# LANGUAGE Strict #-}

module Loops where

import Begin
import qualified Data.Vector.Unboxed.Mutable as MV

data Relations s = Relations
   { ends :: STVector s Int
   , sets :: STRef s Int
   }

create :: Int -> ST s (Relations s)
create n = do
   ends <- MV.generate n id
   sets <- newSTRef 0
   return Relations{ends, sets}

reset :: Relations s -> Int -> ST s ()
reset Relations{ends, sets} n = do
   forM_ [0 .. n - 1] $ \i -> MV.write ends i i
   writeSTRef sets 0

unite :: Relations s -> Int -> Int -> ST s ()
unite Relations{ends, sets} a b = do
   ea <- MV.read ends a
   eb <- MV.read ends b

   if ea == b
      then sets += 1
      else do
         MV.write ends ea eb
         MV.write ends eb ea

count :: Relations s -> ST s Int
count Relations{sets} = readSTRef sets

File: source/nim/src/loops.nim
# Loops algorithm for cycle counting

import std/sequtils

type
  Loops* = object
    ends: seq[int]
    sets: int

proc create*(n: int): Loops =
  result.ends = toSeq(0..<n)
  result.sets = 0

proc reset*(rel: var Loops, n: int) =
  for i in 0..<n:
    rel.ends[i] = i
  rel.sets = 0

proc unite*(rel: var Loops, a, b: int) =
  let ea = rel.ends[a]
  let eb = rel.ends[b]
  
  if ea == b:
    inc rel.sets
  else:
    rel.ends[ea] = eb
    rel.ends[eb] = ea

proc setCount*(rel: Loops): int =
  rel.sets

const name* = "Loops"
